File: src/actions/doctor.ts
/**
 * Prerequisite checks — reports config, CLI tools, and auth status.
 */
import { join } from "node:path";
import { match } from "ts-pattern";
import { checkAuth, checkAvailable } from "../apis/github/client.ts";
import { type Config, classroomStatus, configFilePath, hasCanvas } from "./config.ts";

export type CheckStatus = "ok" | "warn" | "error";

export interface Check {
	name: string;
	status: CheckStatus;
	detail: string;
	indent: number;
}

function ok(name: string, detail: string, indent = 0): Check {
	return { name, status: "ok", detail, indent };
}

function warn(name: string, detail: string, indent = 0): Check {
	return { name, status: "warn", detail, indent };
}

function err(name: string, detail: string, indent = 0): Check {
	return { name, status: "error", detail, indent };
}

/** Run all prerequisite checks and return results. */
export async function checkPrerequisites(cfg?: Config): Promise<Check[]> {
	const checks: Check[] = [];

	// 1. Config file
	const path = await configFilePath();
	if (!path) {
		checks.push(err("cass.toml", "not found in any parent directory"));
		return checks;
	}
	checks.push(ok("cass.toml", path));

	if (!cfg) return checks;

	// 2. Classroom checks
	const state = classroomStatus(cfg);
	await match(state)
		.with({ tag: "configured" }, async (s) => {
			checks.push(ok("classroom", `gh_id=${s.ghId}`, 1));
			await addGHChecks(checks);
		})
		.with({ tag: "pending" }, async (s) => {
			checks.push(warn("classroom", `url_id=${s.urlId} (gh_id unresolved)`, 1));
			await addGHChecks(checks);
		})
		.with({ tag: "missing" }, () => {
			checks.push(warn("classroom", "not configured (optional)", 1));
		})
		.exhaustive();

	// 3. Canvas checks
	if (hasCanvas(cfg)) {
		checks.push(ok("canvas", cfg.canvasBaseUrl, 1));
		checks.push(ok("course_id", String(cfg.canvasCourseId), 2));

		// Check token
		const tokenPath = join(cfg.root, ".canvastoken");
		const tokenExists = await Bun.file(tokenPath).exists();
		const envToken = Boolean(process.env.CANVAS_TOKEN);

		if (tokenExists) {
			checks.push(ok("canvas token", ".canvastoken", 2));
		} else if (envToken) {
			checks.push(ok("canvas token", "$CANVAS_TOKEN", 2));
		} else {
			checks.push(err("canvas token", "not found (.canvastoken or $CANVAS_TOKEN)", 2));
		}
	} else {
		checks.push(warn("canvas", "not configured", 1));
	}

	return checks;
}

async function addGHChecks(checks: Check[]): Promise<void> {
	const ghAvailable = await checkAvailable();
	if (!ghAvailable) {
		checks.push(err("gh CLI", "not found", 2));
		return;
	}
	checks.push(ok("gh CLI", "available", 2));

	const [authed, username] = await checkAuth();
	if (authed) {
		checks.push(ok("gh auth", username, 2));
	} else {
		checks.push(err("gh auth", "not authenticated", 2));
	}
}

File: cass/actions/doctor.py
"""Prerequisite checks for cass setup (``cass init`` doctor)."""

__docformat__ = "google"


import shutil
from dataclasses import dataclass

from ..apis.canvas.auth import find_auth
from .config import config_file_path, get_config


@dataclass
class Check:
    name: str
    ok: bool
    detail: str
    indent: int = 0


def check_prerequisites() -> list[Check]:
    """Run all setup checks and return results."""
    from ..apis.github.client import check_auth, check_available
    from ..apis.github.service import check_classroom_extension

    checks: list[Check] = []

    # Config file
    cfg_path = config_file_path()
    if not cfg_path:
        checks.append(Check("cass.toml", False, "not found — run 'cass init'"))
        return checks

    checks.append(Check("cass.toml", True, f"found at {cfg_path}"))

    try:
        cfg = get_config()
    except SystemExit:
        checks.append(Check("config", False, "invalid — needs a [canvas] section"))
        return checks

    # GitHub Classroom checks (only if configured)
    if cfg.has_classroom:
        checks.append(Check("classroom", True, "configured"))
        checks.append(
            Check(
                "classroom_url",
                cfg.has_classroom_url,
                cfg.classroom_url or "not set",
                indent=1,
            )
        )
        checks.append(
            Check(
                "url_id", cfg.classroom_url_id > 0, str(cfg.classroom_url_id), indent=1
            )
        )
        checks.append(
            Check("gh_id", cfg.classroom_gh_id > 0, str(cfg.classroom_gh_id), indent=1)
        )

        # git
        checks.append(
            Check(
                "git",
                bool(shutil.which("git")),
                "installed" if shutil.which("git") else "not found",
            )
        )

        # gh CLI
        gh_ok = check_available()
        checks.append(
            Check(
                "gh",
                gh_ok,
                "installed" if gh_ok else "not found — https://cli.github.com/",
            )
        )
        if gh_ok:
            ext_ok, ext_detail = check_classroom_extension()
            checks.append(Check("gh classroom", ext_ok, ext_detail, indent=1))
            authed, detail = check_auth()
            checks.append(Check("gh auth", authed, detail, indent=1))
    elif cfg.classroom_needs_resolution:
        checks.append(
            Check("classroom", False, "URL saved; gh-classroom ID unresolved")
        )
        checks.append(Check("classroom_url", True, cfg.classroom_url, indent=1))
        checks.append(Check("url_id", True, str(cfg.classroom_url_id), indent=1))
        checks.append(Check("gh_id", False, "unresolved", indent=1))
        gh_ok = check_available()
        checks.append(
            Check(
                "gh",
                gh_ok,
                "installed" if gh_ok else "Install GitHub CLI: https://cli.github.com/",
            )
        )
        if gh_ok:
            ext_ok, ext_detail = check_classroom_extension()
            checks.append(Check("gh classroom", ext_ok, ext_detail, indent=1))
            authed, detail = check_auth()
            checks.append(Check("gh auth", authed, detail, indent=1))
    else:
        checks.append(Check("classroom", True, "not configured (optional)"))

    # Canvas checks (only if configured)
    if cfg.has_canvas:
        checks.append(Check("canvas", True, "configured"))
        checks.append(Check("base_url", True, cfg.canvas_base_url, indent=1))
        checks.append(Check("course_id", True, str(cfg.canvas_course_id), indent=1))
        auth = find_auth(cfg.root)
        checks.append(
            Check(
                "auth",
                auth is not None,
                auth.description
                if auth
                else "not found (.canvascreds, .canvastoken or $CANVAS_TOKEN)",
                indent=1,
            )
        )
    else:
        checks.append(Check("canvas", False, "not configured"))

    # Roster
    try:
        from .. import db

        if db.students_exist():
            students = db.load_students()
            checks.append(
                Check("roster", True, f"{len(students)} students in database")
            )
        else:
            checks.append(Check("roster", False, "empty — run 'cass pull'"))
    except SystemExit:
        pass

    return checks
